File: src/org/newtonpolyhedron/solverprinters/MatrixDetSolverPrinter.java
package org.newtonpolyhedron.solverprinters;

import java.io.PrintWriter;

import org.apache.commons.math3.fraction.BigFraction;
import org.apache.commons.math3.linear.FieldMatrix;
import org.newtonpolyhedron.entity.SolverPrinter;
import org.newtonpolyhedron.utils.MatrixUtils;

public class MatrixDetSolverPrinter extends SolverPrinter <Void> {
	
	private final FieldMatrix <BigFraction>	baseMatrix;
	private final int						skipRow;
	private final int						skipCol;
	
	public MatrixDetSolverPrinter(
			final FieldMatrix <BigFraction> baseMatrix,
			final int skipRow,
			final int skipCol,
			final PrintWriter output) {
		super(null, output);
		this.baseMatrix = baseMatrix;
		this.skipRow = skipRow;
		this.skipCol = skipCol;
	}
	
	@Override
	public void solveFor(final Void nothing, final PrintWriter output) {
		final BigFraction det = MatrixUtils.getDet(baseMatrix, skipRow, skipCol);
		output.println(subheader("Base matrix:"));
		output.println(MatrixUtils.toString(baseMatrix));
		output.println("Matrix determinant: " + det);
	}
}

File: src/org/newtonpolyhedron/solverprinters/MatrixDetSolverPrinter.scala
package org.newtonpolyhedron.solverprinters
import java.io.PrintWriter

import org.apache.commons.math3.fraction.BigFraction
import org.apache.commons.math3.linear.FieldMatrix
import org.newtonpolyhedron.entity.BigFrac
import org.newtonpolyhedron.entity.Matrix
import org.newtonpolyhedron.entity.SolverPrinter

class MatrixDetSolverPrinter(val baseMatrix: Matrix[BigFrac],
                             val skipRow: Int,
                             val skipCol: Int,
                             output: PrintWriter)
    extends SolverPrinter[Void](null, output) {

  def this(baseMatrix: FieldMatrix[BigFraction],
           skipRow: Int,
           skipCol: Int,
           output: PrintWriter) =
    this(Matrix(baseMatrix.getData map (_ map (BigFrac(_)))), skipRow, skipCol, output)

  override def solveFor(nothing: Void,
                        output: PrintWriter) = {
    val det = baseMatrix.minor(skipRow, skipCol)
    output.println(subheader("Base matrix:"))
    output.println(baseMatrix)
    if (skipRow != -1 || skipCol != -1) {
      (skipRow, skipCol) match {
        case (-1, -1) => // NOOP
        case (r, -1)  => output.println("(Skipping row " + skipRow + ")")
        case (-1, c)  => output.println("(Skipping col " + skipCol + ")")
        case (r, c)   => output.println("(Skipping row " + skipRow + ", col " + skipCol + ")")
      }
    }
    output.println("Matrix determinant: " + det)
  }
}